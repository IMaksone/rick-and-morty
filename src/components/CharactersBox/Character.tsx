import useMyContext from "src/context/useMyContext";
import { CharacterStatusType, LocalCharacterType } from "src/types/character";
import CharacterWrapper from "./CharacterWrapper";

interface CharacterInterface {
  character: LocalCharacterType;
}

const getIndicatorClassName = (characterStatus: CharacterStatusType) =>
  characterStatus === "Alive"
    ? "character-indicator-alive"
    : "character-indicator-dead";

export const Character = ({ character }: CharacterInterface) => {
  const { index, data } = character;

  const { setModal } = useMyContext();

  const addModal = () => {
    setModal(data);
  };

  const characterStatusClassName =
    "character-text text-dark character-indicator " +
    getIndicatorClassName(data.status);

  return (
    <CharacterWrapper characterIndex={index}>
      <div className="character">
        <img
          alt="character_image"
          className="character-img"
          src={data.image}
          onClick={addModal}
        />
        <div className="character-desc">
          <div>
            <h3 className="character-name text-light" onClick={addModal}>
              {data.name}
            </h3>
            <p className={characterStatusClassName}>
              {data.status.toLowerCase()}
            </p>
          </div>
          <p className="character-text text-dark">{data.species}</p>
        </div>
      </div>
    </CharacterWrapper>
  );
};
