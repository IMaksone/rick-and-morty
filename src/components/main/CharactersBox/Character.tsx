import { charactersBoxParams } from "src/constants";
import { useMyContext } from "src/context";
import { LocalCharacterType } from "src/types/character";

interface CharacterInterface {
  character: LocalCharacterType;
}

export const Character = ({ character }: CharacterInterface) => {
  const { setModal } = useMyContext();

  const addModal = () => {
    setModal(character.data);
  };

  return (
    <div
      className="character-wr"
      style={{
        height: charactersBoxParams.elHeight,
        top: character.index * charactersBoxParams.elFullHeight
      }}
    >
      <div className="character">
        <img
          alt="character_image"
          className="character-img"
          src={character.data.image}
          onClick={addModal}
        />
        <div className="character-desc">
          <div>
            <h3 className="character-name text-light" onClick={addModal}>
              {character.data.name}
            </h3>
            <p
              className={`character-text text-dark character-indicator ${
                character.data.status === "Alive"
                  ? "character-indicator-alive"
                  : "character-indicator-dead"
              }`}
            >
              {character.data.status.toLowerCase()}
            </p>
          </div>
          <p className="character-text text-dark">{character.data.species}</p>
        </div>
      </div>
    </div>
  );
};
