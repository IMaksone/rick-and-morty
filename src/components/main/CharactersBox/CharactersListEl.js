import { charactersBoxParams } from "../../../constants";
import { useMyContext } from "../../../context";
import { myCharacterType } from "../../../types";

export const CharactersListEl = ({ character }) => {
  const { setModal } = useMyContext();

  const addModal = () => {
    setModal(character.data);
  };

  return (
    <div
      className="characters-list-el"
      style={{
        height: charactersBoxParams.elHeight,
        top: character.index * charactersBoxParams.elFullHeight,
      }}
    >
      <div className="character">
        <img
          alt="character_image"
          className="character-img"
          src={character.data.image}
          onClick={addModal}
        />
        <div className="character-desc">
          <div>
            <h3 className="character-name text-light" onClick={addModal}>
              {character.data.name}
            </h3>
            <p
              className={`character-text text-dark character-indicator ${
                character.data.status === "Alive"
                  ? "character-indicator-alive"
                  : "character-indicator-dead"
              }`}
            >
              {character.data.status.toLowerCase()}
            </p>
          </div>
          <p className="character-text text-dark">{character.data.species}</p>
        </div>
      </div>
    </div>
  );
};

CharactersListEl.propTypes = {
  character: myCharacterType.isRequired,
};
