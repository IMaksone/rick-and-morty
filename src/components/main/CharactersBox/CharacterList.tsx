import { Character } from "./Character";
import { charactersBoxParams } from "src/constants";
import { LocalCharacterType } from "src/types/character";

interface CharacterListInterface {
  characterList: LocalCharacterType[],
  count: number
}

export const CharacterList = ({ characterList, count }: CharacterListInterface) => {
  return (
    <div
      className="character-list"
      style={{
        height: charactersBoxParams.elHeight * count,
      }}
    >
      {characterList.map((ch, i) => (
        <Character key={i} character={ch} />
      ))}
    </div>
  );
};

