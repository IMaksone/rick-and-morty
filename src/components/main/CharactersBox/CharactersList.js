import PropTypes from "prop-types";

import { CharactersListEl } from "./CharactersListEl";
import { charactersBoxParams } from "../../../constants";
import { myCharacterType } from "../../../types/character";

export const CharactersList = ({ render, length }) => {
  return (
    <div
      className="characters-list"
      style={{
        height: charactersBoxParams.elHeight * length,
      }}
    >
      {render.map((ch, i) => (
        <CharactersListEl key={i} character={ch} />
      ))}
    </div>
  );
};

CharactersList.propTypes = {
  render: PropTypes.arrayOf(myCharacterType).isRequired,
  length: PropTypes.number.isRequired,
};

