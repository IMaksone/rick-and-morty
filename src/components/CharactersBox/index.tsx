import { CharacterList } from "./CharacterList";
import useCharactersBox from "./useCharactersBox";

import "./charactersBox.css";

export const CharactersBox = () => {
  const { ref, visibleCharacters, filteredCharactersLength, handleScroll } =
    useCharactersBox();

  return (
    <div className="characters-box" ref={ref} onScroll={handleScroll}>
      <CharacterList
        visibleCharacters={visibleCharacters}
        count={filteredCharactersLength}
      />
    </div>
  );
};
