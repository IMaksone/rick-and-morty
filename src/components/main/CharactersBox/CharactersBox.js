import { useEffect, useRef, useState } from "react";

import { useMyContext } from "../../../context";
import { CharactersList } from "./CharacterList";
import { charactersPrepare, renderPreparation } from "../../../scripts";

import "./charactersBox.css";

export const CharactersBox = () => {
  const { characters, filters } = useMyContext();

  const ref = useRef({ current: undefined });

  const [prepared, setPrepared] = useState([]);
  const [render, setRender] = useState([]);

  useEffect(() => {
    setRender(renderPreparation(prepared, ref.current));
  }, [prepared]);

  useEffect(() => {
    setPrepared(charactersPrepare(characters, filters));
  }, [characters, filters]);

  const handleScroll = (e) => {
    setRender(renderPreparation(prepared, e.target));
  };

  return (
    <div className="characters-box" ref={ref} onScroll={handleScroll}>
      <CharactersList render={render} length={prepared.length} />
    </div>
  );
};
