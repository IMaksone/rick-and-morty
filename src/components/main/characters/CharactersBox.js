import { useEffect, useRef, useState } from "react";
import { useMyContext } from "../../../context";

import { CharactersList } from "./CharactersList";
import { charactersPrepare, renderPreparation } from "../../../scripts";

import "./charactersBox.css";

export const CharactersBox = () => {
  const { characters, filters } = useMyContext();

  const listRef = useRef({ current: undefined });

  const [prepared, setPrepared] = useState(
    charactersPrepare(characters, filters)
  );
  const [render, setRender] = useState(prepared);

  useEffect(() => {
    setRender(renderPreparation(prepared, listRef.current));
  }, [prepared]);

  useEffect(() => {
    setPrepared(charactersPrepare(characters, filters));
  }, [characters, filters]);

  const handleScroll = (e) => {
    setRender(renderPreparation(prepared, e.target));
  };

  return (
    <div className="characters-box" onScroll={handleScroll}>
      <CharactersList
        listRef={listRef}
        render={render}
        length={prepared.length}
      />
    </div>
  );
};
