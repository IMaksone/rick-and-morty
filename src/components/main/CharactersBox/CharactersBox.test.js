import renderer from "react-test-renderer";

import { CharactersBox } from "./CharactersBox";

it("TEST CharactersBox", () => {
  const component = renderer.create(<CharactersBox />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
