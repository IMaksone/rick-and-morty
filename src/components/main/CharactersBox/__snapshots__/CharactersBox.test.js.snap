// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST CharactersBox 1`] = `
<div
  className="characters-box"
  onScroll={[Function]}
>
  <div
    className="characters-list"
    style={
      Object {
        "height": 0,
      }
    }
  />
</div>
`;
